README.md
basic_vsftpd Cookbook
=====================
[![Build Status](https://travis-ci.org/flatrocks/cookbook-basic_vsftpd.svg?branch=master)](https://travis-ci.org/flatrocks/cookbook-basic_vsftpd)

A basic cookbook to install and configure the vsftpd package.

All vsftpd configuration options can be set using this cookbook,
and no "helpful" defaults are provided.
See [Attributes](#attributes) for details.

## Supported platforms
This cookbook has only been tested for Ubuntu 14.04.

Testing/adding support for other platforms is encouraged.  See [Contributing](#contributing).

## Recipes

### default
This recipe:
* installs the vsftpd package
* renames the original vsftpd.conf file to vsftpd.conf.orig for reference
* creates a new vsftpd.conf configuration file
from ```node['basic_vsftpd']['config']['...']``` attributes
* if ```node['basic_vsftpd']['config']['user_config_dir']``` is set, creates an empty user config directory
* enables and starts the ```vsftpd``` service

### chroot_list
Creates a chroot_list file for vsftpd from node attributes.
See [Other attributes](@other-attributes) for details.
_If you don't want to use this recipe, you can create the chroot_list file manually._

### userlist
Creates a userlist file for vsftpd from node attributes.
See [Other attributes](@other-attributes) for details.
_If you don't want to use this recipe, you can create the userlist file manually._

## Resources

### user_conf
This resource creates a user configuration file in the vsftpd user_config_dir directory.

Properties:
* __user__, string, name property, the name of the user
* __content__, string, the content of the user config file

Actions:
* :create, :delete

Note: This resource requires the ```node['basic_vsftpd']['config']['user_config_dir']```
attribute to be set and for the directory to exist.

Attributes
----------

### vsftpd config attributes
Only two vsftpd configuration attributes are set by default.
The default values are the same as the vsftpd defaults,
and are set only to make the defaults available to the node.
* __node['basic_vsftpd']['config']['chroot_list_file']__, default: '/etc/vsftpd.chroot_list'
* __node['basic_vsftpd']['config']['userlist_file']__, default: '/etc/vsftpd.user_list'

To set other _vsftpd.conf_ attributes
```
# string attributes:
default['basic_vsftpd']['config']['option_name'] = 'string_option_value'
# YES/NO attributes
default['basic_vsftpd']['config']['yesno_option_name'] = 'YES' # or 'NO'
# numeric attributes
default['basic_vsftpd']['config']['numeric_option_name'] = 12345
default['basic_vsftpd']['config']['numeric_option_name'] = '12345' # also OK
```

Note: When setting ```node['basic_vsftpd']['config'][...]``` attributes using Ruby,
each should be set individually, __not__ all at once using a Ruby hash.  That is,
```
# Yes
default['basic_vsftpd']['config']['option1'] = 'setting1'
default['basic_vsftpd']['config']['option2'] = 'setting2'

# NO!
default['basic_vsftpd']['config'] = {
  'option1' => 'setting1',
  'option2' => 'setting2'
}
```

### Other attributes
* __node['basic_vsftpd']['chroot_list']__, default: [ ], used by the chroot_list recipe, list of
user names to fill vsftpd's chroot_list file.
* __node['basic_vsftpd']['userlist']__, default: [ ], used by the userlist recipe, list of
user names to fill vsftpd's userlist file.

Requirements
------------


Contributing
------------

__Examples:__

Often the real trick is in configuring vsftpd,
for a specific application,
but that's outside the scope of this cookbook.
Please consider documenting useful configuration examples
by adding a page to the wiki.

__Code:__

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Author: Tom Wilson (tom at rollnorocks.com)

    Copyright (C) 2015, Tom Wilson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


CHANGELOG.md
basic_vsftpd CHANGELOG
======================

0.1.4
-----
- README update only

0.1.3
-----
- Keep original vsftpd.conf file on system for reference

0.1.2
-----
- Updates to metadata comments for supermarket

0.1.1
-----
- Added Travis CI support
- Cosmetic changes to license, headers, and readme.

0.1.0
-----
- Initial release of basic_vsftpd

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

RSpec.configure do |config|
  # Specify the operating platform to mock Ohai data from
  config.platform = 'ubuntu'

  # Specify the operating version to mock Ohai data from
  config.version = '14.04'
end

ChefSpec::Coverage.start!


spec/recipes/default_spec.rb
# encoding: utf-8
require 'spec_helper'

describe 'basic_vsftpd::default' do
  subject { ChefSpec::SoloRunner.new() do |node|
    node.set['basic_vsftpd']['config']['a_fake_key'] = 'a_fake_value'
  end.converge(described_recipe) }

  it 'installs vsftpd' do
    expect(subject).to install_package('vsftpd')
  end

  describe 'the vsftpd.conf file' do
    let(:config_file) {'/etc/vsftpd.conf'}
    let(:config_orig) {'/etc/vsftpd.conf.orig'}
    it 'original contents are copied' do
      expect(subject).to create_remote_file_if_missing(config_orig)
    end
    it 'is created' do
      expect(subject).to create_template(config_file)
    end
    it 'has the right mode, group and owner' do
      expect(subject).to create_template(config_file).with_owner('root').with_group('root').with_mode('0400')
    end
    it 'has the expected content' do
      [ 'a_fake_key=a_fake_value',
        'chroot_list_file=/etc/vsftpd.chroot_list',
        'userlist_file=/etc/vsftpd.user_list'
      ].each do |expected_content|
        expect(subject).to render_file(config_file).with_content expected_content
      end
    end
  end

  describe 'the user config directory' do
    context 'when vsftpd user_config_dir is set' do
      let(:the_dir) {'/some/dir'}
      subject { ChefSpec::SoloRunner.new() do |node|
        node.set['basic_vsftpd']['config']['user_config_dir'] = the_dir
      end.converge(described_recipe) }

      it 'is created' do
        expect(subject).to create_directory(the_dir)
      end
      it 'has the right mode, group and owner' do
        expect(subject).to create_directory(the_dir).with_owner('root').with_group('root').with_mode('0500')
      end
    end
    context 'when vsftpd user_config_dir is not set' do
      it 'is not created' do
        # nothing to create, but don't raise error
        expect{subject}.to_not raise_error
      end
    end
  end

  describe 'the vsftpd service' do
    it 'is enabled and started' do
      expect(subject).to enable_service('vsftpd')
      expect(subject).to start_service('vsftpd')
    end
    it 'has the right provider' do
      expect(subject).to enable_service('vsftpd').with_provider Chef::Provider::Service::Upstart
      expect(subject).to start_service('vsftpd').with_provider Chef::Provider::Service::Upstart
    end
  end

end





spec/recipes/userlist_spec.rb
# encoding: utf-8
require 'spec_helper'

describe 'basic_vsftpd::userlist' do
  subject { ChefSpec::SoloRunner.new() do |node|
  end.converge(described_recipe) }

  context 'with default userlist file' do
    subject { ChefSpec::SoloRunner.new() do |node|
    end.converge(described_recipe) }
    describe 'the userlist_file' do
      describe 'its directory' do
        it 'is created recursively' do
          allow(File).to receive(:exists?).and_call_original
          allow(File).to receive(:exists?).with('/etc') {false}
          expect(subject).to create_directory('/etc')
        end
        it 'is not created if exists' do
          allow(File).to receive(:exists?).and_call_original
          allow(File).to receive(:exists?).with('/etc') {true}
          expect(subject).to_not create_directory('/etc')
        end
      end
      it 'is created' do
        expect(subject).to create_file('/etc/vsftpd.user_list')
      end
    end
  end

  context 'with custom userlist file' do
    describe 'the userlist_file' do
      let(:file_dir) {'/some/dir'}
      let(:the_file) {file_dir + '/the_users'}
      subject { ChefSpec::SoloRunner.new() do |node|
        node.set['basic_vsftpd']['config']['userlist_file'] = the_file
        node.set['basic_vsftpd']['userlist'] = ['user1', 'user2']
      end.converge(described_recipe) }
      describe 'its directory' do
        it 'is created recursively' do
          expect(subject).to create_directory(file_dir).with_recursive(true)
        end
        it 'has the right mode, group and owner' do
          expect(subject).to create_directory(file_dir).with_owner('root').with_group('root').with_mode('0500')
        end
      end
      it 'is created' do
        expect(subject).to create_file(the_file)
      end
      it 'has the right mode, group and owner' do
        expect(subject).to create_file(the_file).with_owner('root').with_group('root').with_mode('0400')
      end
      it 'has the expected content' do
        expect(subject).to render_file(the_file).with_content "user1\nuser2"
      end
    end
  end

end





recipes/default.rb
#
# Cookbook Name:: basic_vsftpd
# Recipe:: default
#
# Copyright (C) 2015, Tom Wilson
#
# All rights reserved - Do Not Redistribute
#

package 'vsftpd'

vsftpd_config_file = value_for_platform_family(
  'rhel'   => '/etc/vsftpd/vsftpd.conf',
  'debian' => '/etc/vsftpd.conf',
  'default' => '/etc/vsftpd.conf'
)

remote_file 'Copy original vsftpd config' do
  path "#{vsftpd_config_file}.orig"
  source "file://#{vsftpd_config_file}"
  owner 'root'
  group 'root'
  mode '0400'
  action :create_if_missing
end

template vsftpd_config_file do
  source 'vsftpd.conf.erb'
  variables options: node['basic_vsftpd']['config']
  owner 'root'
  group 'root'
  mode '0400'
end

if node['basic_vsftpd']['config']['user_config_dir']
  directory node['basic_vsftpd']['config']['user_config_dir'] do
    recursive true
    owner 'root'
    group 'root'
    mode '0500'
  end
end

service 'vsftpd' do
  provider Chef::Provider::Service::Upstart
  action [:enable, :start]
  supports restart: true
end
